src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -285,16 +263,127 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                 context.Write(writer, new DynamicJsonValue
                 {
                     [nameof(BatchCommandResult.Results)] = array,
-                    [nameof(BatchCommandResult.TransactionIndex)] = result.Index
+                    [nameof(BatchCommandResult.TransactionIndex)] = index
                 });
             }
         }
 
+        private async Task<long?> GetClusterTransactionCount(object result, long index, long databaseCommandsCount, TaskCompletionSource<long?> onDatabaseCompletionTcs)
+        {
+            if (databaseCommandsCount == 0)
+                return null;
+
+            long? count = null;
+
+            if (result is List<ClusterTransactionCommand.ClusterTransactionErrorInfo> errors)
+                ThrowClusterTransactionConcurrencyException(errors);
+
+            if (result != null && result is ClusterTransactionResult clusterTxResult)
+            {
+                if(clusterTxResult.Count.HasValue == false)
+                    throw new InvalidOperationException($"Cluster Transaction result is null, but has more then 0 database commands.");
+
+                count = clusterTxResult.Count.Value;
+            
+                Database.ForTestingPurposes?.AfterCommitInClusterTransaction?.Invoke();
+            
+                var token = HttpContext.RequestAborted;
+                await using (token.Register(() => onDatabaseCompletionTcs.TrySetCanceled()))
+                    await onDatabaseCompletionTcs.Task.ConfigureAwait(false);
+                
+            }
+            else // leader isn't updated
+            {
+                // searching in the history log
+                var token = HttpContext.RequestAborted;
+                await using (token.Register(() => onDatabaseCompletionTcs.TrySetCanceled()))
+                    count = await onDatabaseCompletionTcs.Task.ConfigureAwait(false);  // we'll try to take the 'count' from the local task that we created for this ClusterTransactionCommand
+            
+                if (count.HasValue == false) // Failover and the ClusterTransactionCommand is already completed in the DocumentDatabase, so the task with the result is no longer exists, so we'll try to take the result from the local history log.
+                {
+                    using (ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+                    using (ctx.OpenReadTransaction())
+                    {
+                        if (ServerStore.Engine.LogHistory.TryGetCommandByIndex(ctx, index, out var localCommand))
+                        {
+                            var clusterTxLocalResult = JsonConvert.DeserializeObject<ClusterTransactionResult>(localCommand["Result"].ToString());
+                            if (clusterTxLocalResult.Count.HasValue == false) //clusterTxLocalResult.Count == null and databaseCommandsCount>0 => clusterTxLocalResult.Count is 0 (DeserializeObject convert 0 to null)
```

Comment:
Are you sure?
```
[Fact]
public void TestCase()
{
    string str = @"{
""Count"": 0
}";
    var r = JsonConvert.DeserializeObject<ClusterTransactionResult>(str);
    Assert.NotNull(r);
}
```

Reply:
![image](https://github.com/ravendb/ravendb/assets/48368355/64622d29-e3f8-4204-ab1d-92a7a4101489)
![image](https://github.com/ravendb/ravendb/assets/48368355/ecc39208-7a69-4b9c-b845-9d059148c392)

Reply:
The `ClusterTransactionResult` is a nested object in the serialized object (the `Result` property).
You cannot deserialize it like that.
You got `null` because the serialized object doesn't have the `Count` property

Reply:
Done